src/workerd/api/sockets.h

```diff
@@ -102,8 +43,26 @@ class Socket: public jsg::Object {
 private:
   jsg::Ref<ReadableStream> readable;
   jsg::Ref<WritableStream> writable;
+  kj::Own<jsg::PromiseResolverPair<void>> closeFulfiller;
+  // This fulfiller is used to resolve the `closedPromise` below.
+  jsg::MemoizedIdentity<jsg::Promise<void>> closedPromise;
+
   kj::Promise<kj::Own<kj::AsyncIoStream>> processConnection();
-  IoOwn<kj::PromiseFulfillerPair<void>> closeFulfiller;
+
+  void resolveFulfiller(kj::Maybe<kj::Exception> maybeErr) {
+    auto& context = IoContext::current();
+    KJ_IF_MAYBE(err, maybeErr) {
+      closeFulfiller->resolver.reject(context.getCurrentLock(), kj::cp(*err));
```

Comment:
I'd suggest making `jsg::Lock&` be a parameter to `resolveFulfiller()` so that you don't have to look up the current `IoContext`. Both call sites have a `jsg::Lock` readily available.

`context.getCurrentLock()` is kind of a hack that I'd like to remove eventually.